File: examples/sdk-typescript/create-destination.ts
import inquirer from 'inquirer';
import dotenv from "dotenv";
dotenv.config();
import { Outpost } from "@hookdeck/outpost-sdk";
import { CreateTenantDestinationRequest } from '@hookdeck/outpost-sdk/dist/esm/models/operations';

const ADMIN_API_KEY = process.env.ADMIN_API_KEY;
const TENANT_ID = process.env.TENANT_ID || "hookdeck";
const SERVER_URL = process.env.OUTPOST_URL || "http://localhost:3333";

if (!ADMIN_API_KEY) {
  console.error("Please set the ADMIN_API_KEY environment variable.");
  process.exit(1);
}

async function main() {
  const outpostAdmin = new Outpost({
    security: { adminApiKey: ADMIN_API_KEY },
    serverURL: `${SERVER_URL}/api/v1`,
  });

  await outpostAdmin.tenants.upsert({
    tenantId: TENANT_ID,
  });

  const { destinationType } = await inquirer.prompt([
    {
      type: 'list',
      name: 'destinationType',
      message: 'Select a destination type:',
      choices: ['azure_servicebus'],
    },
  ]);

  let destinationCreateRequest: CreateTenantDestinationRequest | null = null;

  if (destinationType === 'azure_servicebus') {
    console.log(`
You can create a topic or queue specific connection string with send-only permissions using the Azure CLI.
Please replace $RESOURCE_GROUP, $NAMESPACE_NAME, and $TOPIC_NAME with your actual values.

Create a send-only policy for the topic:
\`\`\`bash
az servicebus topic authorization-rule create \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --topic-name $TOPIC_NAME \\
  --name SendOnlyPolicy \\
  --rights Send
\`\`\`

or for a queue:

\`\`\`bash
az servicebus queue authorization-rule create \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --queue-name $QUEUE_NAME \\
  --name SendOnlyPolicy \\
  --rights Send
\`\`\`

Get the Topic-Specific Connection String:
\`\`\`bash
az servicebus topic authorization-rule keys list \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --topic-name $TOPIC_NAME \\
  --name SendOnlyPolicy \\
  --query primaryConnectionString \\
  --output tsv
\`\`\`

or for a Queue-Specific Connection String:
\`\`\`bash
az servicebus queue authorization-rule keys list \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --queue-name $QUEUE_NAME \\
  --name SendOnlyPolicy \\
  --query primaryConnectionString \\
  --output tsv
\`\`\`

`);
    const { connectionString, topic_or_queue_name } = await inquirer.prompt([
        {
            type: 'input',
            name: 'connectionString',
            message: 'Enter Azure Service Bus Connection String:',
        },
        {
            type: 'input',
            name: 'topic_or_queue_name',
            message: 'Enter Azure Service Bus Topic or Queue name:',
        }
    ]);
    destinationCreateRequest = {
      tenantId: TENANT_ID,
      destinationCreate: {
        credentials: {
          connectionString,
        },
        type: "azure_servicebus",
        config: {
          name: topic_or_queue_name
        },
        topics: "*",
      }
    }

  } else {
    console.log(`Destination type "${destinationType}" is not supported by this script.`);
    return;
  }

  try {
    const destination = await outpostAdmin.destinations.create(destinationCreateRequest);
    console.log("Destination created successfully:", destination);
  } catch (error) {
    console.error("An error occurred:", error);
  }
}

main();

File: examples/sdk-python/example/create_destination.py
import os
import sys
import questionary
from dotenv import load_dotenv
from outpost_sdk import Outpost, models


def run():
    """
    Interactively creates a new destination for a tenant.
    """
    load_dotenv()

    admin_api_key = os.getenv("ADMIN_API_KEY")
    tenant_id = os.getenv("TENANT_ID", "hookdeck")
    server_url = os.getenv("OUTPOST_URL", "http://localhost:3333")

    if not admin_api_key:
        print("Please set the ADMIN_API_KEY environment variable.")
        sys.exit(1)

    sdk = Outpost(
        security=models.Security(admin_api_key=admin_api_key),
        server_url=f"{server_url}/api/v1",
    )

    sdk.tenants.upsert(tenant_id=tenant_id)

    print(
        """
You can create a topic or queue specific connection string with
send-only permissions using the Azure CLI.
Please replace $RESOURCE_GROUP, $NAMESPACE_NAME, and $TOPIC_NAME
with your actual values.

Create a send-only policy for the topic:
```bash
az servicebus topic authorization-rule create \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --topic-name $TOPIC_NAME \\
  --name SendOnlyPolicy \\
  --rights Send
```

or for a queue:

```bash
az servicebus queue authorization-rule create \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --queue-name $QUEUE_NAME \\
  --name SendOnlyPolicy \\
  --rights Send
```

Get the Topic-Specific Connection String:
```bash
az servicebus topic authorization-rule keys list \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --topic-name $TOPIC_NAME \\
  --name SendOnlyPolicy \\
  --query primaryConnectionString \\
  --output tsv
```

or for a Queue-Specific Connection String:
```bash
az servicebus queue authorization-rule keys list \\
  --resource-group $RESOURCE_GROUP \\
  --namespace-name $NAMESPACE_NAME \\
  --queue-name $QUEUE_NAME \\
  --name SendOnlyPolicy \\
  --query primaryConnectionString \\
  --output tsv
```
"""
    )

    connection_string = questionary.text(
        "Enter Azure Service Bus Connection String:"
    ).ask()
    if connection_string is None:
        sys.exit(0)

    topic_or_queue_name = questionary.text(
        "Enter Azure Service Bus Topic or Queue name:"
    ).ask()
    if topic_or_queue_name is None:
        sys.exit(0)

    try:
        resp = sdk.destinations.create(
            tenant_id=tenant_id,
            destination_create=models.DestinationCreateAzureServiceBus(
                type=(models.DestinationCreateAzureServiceBusType.AZURE_SERVICEBUS),
                credentials=models.AzureServiceBusCredentials(
                    connection_string=connection_string
                ),
                config=models.AzureServiceBusConfig(name=topic_or_queue_name),
                topics="*",
            ),
        )

        if resp:
            print("Destination created successfully:", resp)
        else:
            print("Failed to create destination, response was empty.")

    except Exception as e:
        print(f"An error occurred: {e}")


if __name__ == "__main__":
    run()
